### PR TITLE
[SPARK-30630][ML] Remove numTrees in GBT in 3.0.0

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
Remove ```numTrees``` in GBT in 3.0.0. 

### Why are the changes needed?
Currently, GBT has 
```
  /**
   * Number of trees in ensemble
   */
  @Since("2.0.0")
  val getNumTrees: Int = trees.length
```
and 
```
  /** Number of trees in ensemble */
  val numTrees: Int = trees.length
```
I think we should remove one of them. We deprecated it in 2.4.5 via https://github.com/apache/spark/pull/27352.

### Does this PR introduce any user-facing change?
Yes, remove ```numTrees``` in GBT in 3.0.0

### How was this patch tested?
existing tests

### CODE REVIEW

**Review by srowen — APPROVED** [review]

File: mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala [hunk 1]
```diff
@@ -342,7 +342,12 @@ class GBTClassificationModel private[ml](
     }
   }
 
-  /** Number of trees in ensemble */
+  /**
+   * Number of trees in ensemble
+   *
+   * @deprecated  Use [[getNumTrees]] instead.  This method will be removed in 3.1.0
```
Comment (dongjoon-hyun) [inline mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: Super nit. `instead.  This` -> `instead. This`?

File: mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala [hunk 2]
```diff
@@ -342,7 +342,12 @@ class GBTClassificationModel private[ml](
...
+   */
+  @deprecated("Use getNumTrees instead. This method will be removed in 3.1.0.", "3.0.0")
```
Comment (dongjoon-hyun) [inline mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: I'm wondering if we can deprecate this in 2.4.5 and remove in 3.0.0.

**Comment by dongjoon-hyun** [comment]
@huaxingao . Could you file a JIRA for this?

File: mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala (hunk 2 above)
Comment (huaxingao) [reply mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: so deprecate this in v2.4.5-rc1, right?

**Comment by huaxingao** [comment]
@dongjoon-hyun 

> Could you file a JIRA for this?

Yes, I will open a JIRA. Thank you very much for reviewing this PR.

File: mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: It's `2.4.5` because RC is not an official release.
RC1 vote fails and we are preparing 2.4.5 RC2.

Comment (huaxingao) [reply mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: Got it. Thanks!

**Comment by dongjoon-hyun** [comment]
@huaxingao . We need to deprecate this in `branch-2.4` first. We can not proceed in the reverse order. Could you make a PR for `branch-2.4`?
> I will deprecate it in 2.4.5 and remove it in 3.0.0

**Comment by huaxingao** [comment]
@dongjoon-hyun Yes. I am trying to make a PR for 2.4. I will also fix the MiMa problem.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
cc @zhengruifeng

**Comment by srowen** [comment]
OK by me. I think there was also no big hurry to deprecate in 2.4 to remove in 3.0, but it's pretty trivial either way.

**Comment by dongjoon-hyun** [comment]
Thank you, @srowen .

**Comment by dongjoon-hyun** [comment]
Merged to master. Thank you, @huaxingao and @srowen .

**Comment by huaxingao** [comment]
Thank you very much! @dongjoon-hyun @srowen

**Comment by zhengruifeng** [comment]
Sorry to come late. LGTM

**Comment by dongjoon-hyun** [comment]
Never late~ Thank you, @zhengruifeng . :)

**Comment by gatorsmile** [comment]
@dongjoon-hyun @zhengruifeng @huaxingao @srowen 

We normally do not make such an exception. Can anyone explain how critical it is?

**Comment by gatorsmile** [comment]
Also, I did not even see the migration guide in the PR.

**Comment by srowen** [comment]
Not at all critical, but what's the exception here?

**Comment by huaxingao** [comment]
@gatorsmile It's not critical. It was deprecated in 2.4.5. I will add the release notes.

**Comment by srowen** [comment]
It can't hurt, but I think this is too trivial for a migration guide. It's already in the deprecation message: use method X instead of Y.

**Comment by dongjoon-hyun** [comment]
@gatorsmile . As mentioned in the PR description, we deprecated this at 2.4.5 which will be released before 3.0.0. Do you mean the deprecation in 2.4.5 is also exceptional?

**Comment by gatorsmile** [comment]
Did we deprecate any API in the maintenance releases and then remove the APIs in the next release? Based on my understanding, all the APIs are deprecated in the major releases. Could any of you give me some examples in our previous releases?

**Comment by dongjoon-hyun** [comment]
We don't remove the API at 2.4.6, @gatorsmile . The next of 2.4.5 is 2.4.6.
> Did we deprecate any API in the maintenance releases and then remove the APIs in the next release?

**Comment by gatorsmile** [comment]
I mean the next major release.

**Comment by gatorsmile** [comment]
Also, we should never remove any API in the maintenance release

**Comment by dongjoon-hyun** [comment]
We didn't remove any API in the maintenance release.

**Comment by dongjoon-hyun** [comment]
This is removed at 3.0.0 at the major version, isn't it?

**Comment by dongjoon-hyun** [comment]
Let me rephrase your words~ So, only 2.0.0 or 3.0.0 can do deprecation . Then, 3.0.0 or 4.0.0 can do the removal?

**Comment by gatorsmile** [comment]
Let me make my points clear. Did we deprecate any API in the maintenance releases? 

For example, did we deprecate an API in 2.3.4 then remove the API in 2.4?

**Comment by dongjoon-hyun** [comment]
SPARK-24918 adds `Executor Plugin` at 2.4.0.
SPARK-29399 removed it at 3.0.0.

**Comment by dongjoon-hyun** [comment]
`Delete the incorrect setWeightCol method in LinearSVCModel` at 2.3.4, 2.4.4, 3.0.0?

**Comment by gatorsmile** [comment]
Deprecating the APIs in the major/feature releases, like 2.4.0, 2.3.0, 2.2.0, is common; but deprecating the APIs in the maintenance releases, like 2.2.1, 2.2.2, 2.3.4 is rare.

**Comment by dongjoon-hyun** [comment]
We deleted public API even in minor versions like SPARK-28780.

**Comment by gatorsmile** [comment]
To be honest, this is a big surprise to me. We might need more public discussions for such changes. This will easily break the applications and make our end users frustrated.

**Comment by dongjoon-hyun** [comment]
I totally agree with you. Your guideline also sounds reasonable.

**Comment by gatorsmile** [comment]
Maybe we need some public discussions about the general guideline about the API deprecation/removal in the dev list; and then everyone can follow the same rule in the future? 

I do not want to offend anyone. Will not give an example in the dev-list discussion.

**Comment by dongjoon-hyun** [comment]
Please add the above examples. Otherwise, people don't understand what we did. Additionally, please add the following.

[SPARK-28556][SQL] QueryExecutionListener should also notify Error
- https://github.com/apache/spark/pull/25292/files

This changes the public API at 3.0.0 last July, but we didn't mentioned it at any 2.4.x releases.

**Comment by dongjoon-hyun** [comment]
I don't think it's an offend. (including your arguments). All comments are valid opinions.

**Comment by dongjoon-hyun** [comment]
I gave +1 for the above one and this PR .

**Comment by srowen** [comment]
The policy, such as it exists, has always been roughly to remove deprecated items in a major release (fine of course), or in a minor release after it's been deprecated for at least one prior minor release (aggressive, but can be fine). Yes, we've done both many times. This is the first situation.

@gatorsmile what do you think the rule has been? I'm not sure I have perceived any other disagreement about that.

This particular change is really minor either way. Deprecating is obviously correct. Whether it's removed in 3.0 or 3.1 doesn't matter much, but, why keep it in 3.0? it's obviously an old API oversight.
